hardware_integration/src/mcp2515_can_interface.cpp

```diff
@@ -0,0 +1,383 @@
+		}
+
+		canFrame.dataLength = (buffer[5] & 0x0F);
+		if (isobus::CAN_DATA_LENGTH >= canFrame.dataLength)
+		{
+			if ((read_register(dataRegister, canFrame.data, canFrame.dataLength)) &&
+			    (modify_register(MCPRegister::CANINTF, intfMask, 0)))
+			{
+				retVal = true;
+			}
+		}
+	}
+	return retVal;
+}
+
+bool MCP2515CANInterface::read_frame(isobus::HardwareInterfaceCANFrame &canFrame)
+{
+	bool retVal = false;
+
+	// TODO: implement interrupt driven message reading
+	std::uint8_t status;
+	while (get_read_status(status))
+	{
+		// Check if any of the buffers have a message
+		if (status & 0x01)
+		{
+			retVal = read_frame(canFrame, MCPRegister::RXB0CTRL, MCPRegister::RXB0DATA, 0x01);
+			break;
+		}
+		else if (status & 0x02)
+		{
+			retVal = read_frame(canFrame, MCPRegister::RXB1CTRL, MCPRegister::RXB1DATA, 0x02);
+			break;
+		}
+
+		std::this_thread::sleep_for(std::chrono::milliseconds(RECEIVE_MESSAGE_READ_RATE));
+	}
+
+	return retVal;
+}
+
+bool MCP2515CANInterface::write_frame(const isobus::HardwareInterfaceCANFrame &canFrame,
+                                      const MCPRegister ctrlRegister,
+                                      const MCPRegister sidhRegister)
+{
+	bool retVal = false;
+
+	// Check if the write buffer is empty
+	std::uint8_t ctrl;
+	if (read_register(ctrlRegister, ctrl))
+	{
+		if ((ctrl & 0x08) == 0)
+		{
+			// Buffer is empty now we can write the buffer
+			std::uint8_t buffer[13];
+			if (canFrame.isExtendedFrame)
+			{
+				buffer[3] = static_cast<std::uint8_t>(canFrame.identifier & 0xFF); //< EID0
+				buffer[2] = static_cast<std::uint8_t>((canFrame.identifier >> 8) & 0xFF); //< EID8
+				buffer[1] = static_cast<std::uint8_t>((canFrame.identifier >> 16) & 0x03); //< SIDL EID17-16
+				buffer[1] |= static_cast<std::uint8_t>((canFrame.identifier >> 13) & 0xE0); //< SIDL SID0-2
+				buffer[1] |= 0x08; //< SIDL exide mask
+				buffer[0] = static_cast<std::uint8_t>((canFrame.identifier >> 21) & 0xFF); //< SIDH
+			}
+			else
+			{
+				buffer[1] = static_cast<std::uint8_t>((canFrame.identifier << 5) & 0xE0); // SIDL
+				buffer[0] = static_cast<std::uint8_t>(canFrame.identifier >> 3); // SIDH
+			}
+
+			buffer[4] = canFrame.dataLength; //< DLC
+			memcpy(&buffer[5], canFrame.data, canFrame.dataLength); //< Data
+			if (write_register(sidhRegister, buffer, 5 + canFrame.dataLength))
+			{
+				// Now indicate that the buffer is ready to be sent
+				if (modify_register(ctrlRegister, 0x08, 0x08))
+				{
+					// Now check if the buffer was sent
+					if (read_register(ctrlRegister, ctrl))
+					{
+						if (0 == (ctrl & (0x40 | 0x20 | 0x10)))
+						{
+							retVal = true;
+						}
+					}
+				}
+			}
+		}
+	}
+
+	return retVal;
+}
+
+bool MCP2515CANInterface::write_frame(const isobus::HardwareInterfaceCANFrame &canFrame)
+{
+	bool retVal = false;
+
+	// Try to write the frame to the first buffer
+	retVal = write_frame(canFrame, MCPRegister::TXB0CTRL, MCPRegister::TXB0SIDH);
```

Comment:
Super minor thing, you're assigning false to `retVal` before immediately overwriting it, so you can probably simplify it down to:  `bool retVal = write_frame(canFrame, MCPRegister::TXB0CTRL, MCPRegister::TXB0SIDH);`

Reply:
Simplified to `bool retVal = write_frame(canFrame, MCPRegister::TXB0CTRL, MCPRegister::TXB0SIDH);`